job-info: Support new "success" attribute

chu11 opened the pull request:
Just throwing this up for some initial reactions.

Support a new "success" attribute in the job list service, which simply returns True / False if a job completed successfully (status = 0 on `finished` in the eventlog).  I sort of dislike the attribute name of "success", but I couldn't really think of anything better.  I'll mull over that.

Added a new "success" and "success_hyphen" format option to `flux-jobs`, nothing too exciting
there.  Example output:

```
>flux jobs -A --format="{id} {success}"
JOBID SUCCESS
109734619250688 
10590399496192 True
335712092160 False
293936824320 True
```

```
>flux jobs -A --format="{id} {success_hyphen}"
JOBID SUCCESS
109734619250688 -
10590399496192 True
335712092160 False
293936824320 True
```

The interesting addition, which no idea if I did it well or not, is a new field name called `success_color` (name can be tweaked, trying to think better name).  It'll colorize unsuccessful jobs:

I have no idea how to cut & paste that output into github, so here's a screenshot as an example:

![Screenshot from 2020-03-12 21-13-52](https://user-images.githubusercontent.com/274859/76589284-c58e2f80-64a6-11ea-9154-76af5f76c8d2.png)

I mulled over other ways to do this (command line option?) but somehow this made the most sense to me.  If it was done via command line options, some refactoring will have to be done to have `flux-jobs`  request attributes from `job-info`, but not output them via the "output format" code already there.

SteVwonder commented:
Very nice!  I like the coloring a lot!  The boolean value in the listing is a bit weird to me.  If we could remove the boolean while keeping the color, I'd vote to make it the default format.

I think we'll also probably want to do what `git` and `spack` do: only colorize when printing to a tty and turn off color when piping/redirecting to a file.  Looks like [spack uses `sys.stdout.isatty()` to check for that](https://github.com/spack/spack/blob/4af630308619afb40b1de89d1f7bac0f57b70a38/lib/spack/llnl/util/tty/color.py#L121).

One thought that I'm kind of waffling on, what if color was a separate argument?  `flux jobs --color=status` could color the line based on the success/failure of the job, `--color=state` could have different colors for each job state (strawman: pending is white, running is green, inactive is grey, depend is blue, etc).  The most out there suggestion: `--color=priority` could analyze the priorities in the listing and color the various jobs on a sprectrum based on their individual priority (e.g. highest priority job is white background with black text, lowest priority job is black background with white text, other jobs are shades of grey in between).  Same for runtime and walltime.  (Obviously just spitballing and thinking long term here, not expecting all of that in this PR).

chu11 commented:
> One thought that I'm kind of waffling on, what if color was a separate argument? flux jobs --color=status could color the line based on the success/failure of the job, --color=state could have different colors for each job state (strawman: pending is white, running is green, inactive is grey, depend is blue, etc). The most out there suggestion: --color=priority could analyze the priorities in the listing and color the various jobs on a sprectrum based on their individual priority (e.g. highest priority job is white background with black text, lowest priority job is black background with white text, other jobs are shades of grey in between). Same for runtime and walltime. (Obviously just spitballing and thinking long term here, not expecting all of that in this PR).

Yeah, although I hadn't thought about it much yet, I wasn't exactly sure how to do the color schemes.  The only cmd I'm familiar with that regularly does stdout color is `ls`, and the colors are configured via environment variable.  Which I guess we could emulate that.  In the event anyone wants to colorize in a unique way, that's the most easily comparable thing.

> I think we'll also probably want to do what git and spack do: only colorize when printing to a tty and turn off color when piping/redirecting to a file. Looks like spack uses sys.stdout.isatty() to check for that.

Ahh, thanks!  I think this was about the next thing I as going to lookup on stackoverflow :P

chu11 commented:
> Yeah, although I hadn't thought about it much yet, I wasn't exactly sure how to do the color schemes. The only cmd I'm familiar with that regularly does stdout color is ls, and the colors are configured via environment variable. Which I guess we could emulate that. In the event anyone wants to colorize in a unique way, that's the most easily comparable thing.

Actually, as I think about it more, this is actually a not bad idea (could be via an option as well).  Basically, all of the code in `flux-jobs` can sort of stay the same.  If someone wants to output the format field `success` or `success_hyphen`, that code can stay as is.

Then any colorization options is sort of an independent action.

```
    for job in jobs:
        if color rule specified & job meets color rule;
            sys.stdout.write (the COLOR);
        print(of.format(job))
        sys.stdout.write(RESET);
```

grondo commented:
Nice! I agree with @SteVwonder that `--color` should be a separate option.
Also agree that the color output could be the default. We should figure out a different color for jobs that got exceptions, vs those that just exited with nonzero status. (red vs yellow?). By default should jobs that are INACTIVE+success be green?

Finally, I think the more useful use case for success attribute is to maybe create  a derived state from the low-level job state + success attribute + other job data. In the porcelain tools like `flux jobs`,  I think the INACTIVE state is probably not of much interest to users. Users will want to see some kind of state that indicates the job completed normally (complete?), was canceled (canceled?), exited with nonzero status (failed?), etc.

grondo commented:
>  I wasn't exactly sure how to do the color schemes.

At first I personally wouldn't worry about this. If you put a default color scheme in some kind of structure in the program, eventually `flux-jobs` could be extended to read from a config file, perhaps optionally from user's homedir.

chu11 commented:
> Nice! I agree with @SteVwonder that --color should be a separate option.
Also agree that the color output could be the default. We should figure out a different color for jobs that got exceptions, vs those that just exited with nonzero status. (red vs yellow?). By default should jobs that are INACTIVE+success be green?

Good idea.  But I think the bigger thing is now figuring out how to specify the colors.  I can look to `LS_COLORS` for guidance, but it'll probably be its own unique thing.  I'll open up an issue.  Although I initially thought about putting simple colors into this PR, I'm thinking this + https://github.com/flux-framework/flux-core/issues/2497 should go in first.

grondo commented:
I would say make the colors static for now. Who actually changes LS_COLORS? Another example, `diff` has a `--pallette` option (never seen it used).

chu11 commented:
> Finally, I think the more useful use case for success attribute is to maybe create a derived state from the low-level job state + success attribute + other job data. In the porcelain tools like flux jobs, I think the INACTIVE state is probably not of much interest to users. Users will want to see some kind of state that indicates the job completed normally (complete?), was canceled (canceled?), exited with nonzero status (failed?), etc.

Perhaps some type of "Status" field?  Not so dissimilar from `sacct`'s "State" output?

grondo commented:
> Perhaps some type of "Status" field? Not so dissimilar from sacct's "State" output?

Maybe internally and for formatting the field is called `{derived_state}`, but yeah the heading would just be `STATE`.  (?)

chu11 commented:
Thought, should `success` and `success_hyphen` even be a format specifier option for `flux-jobs`?  Or should `success` only be an attribute uhhh "internally" known?  i.e. used for eventual colorization?

grondo commented:
I don't know I was wondering if it might be useful in scripting, e.g. `flux jobs -o {id}={success}`, but can't really think of a concrete use case. I agree especially `{success_hyphen}` may not be useful in reality... but seems like it doesn't hurt to keep `{success}` for now?

SteVwonder commented:
FWIW, the [`colored`](https://pypi.org/project/colored/) python module looks really handy as we get deeper and deeper into the coloring.  Maybe we can vendor or depend on it at some point to keep from having literal term codes scattered about.

chu11 commented:
> FWIW, the colored python module looks really handy as we get deeper and deeper into the coloring. Maybe we can vendor or depend on it at some point to keep from having literal term codes scattered about.

Yeah, I saw various coloring modules out there but avoided anything that wasn't on TOSS (for the time being).

chu11 commented:
> I don't know I was wondering if it might be useful in scripting, e.g. flux jobs -o {id}={success}, but can't really think of a concrete use case. I agree especially {success_hyphen} may not be useful in reality... but seems like it doesn't hurt to keep {success} for now?

If we keep 1, we can probably just keep both for now.  We can always remove later if we want to.

grondo commented:
> We can always remove later if we want to.

Agreed. But you know if you leave it in for 1 week someone's code will depend on it forever. :stuck_out_tongue:

chu11 commented:
re-pushed, removing the colorizing output (that's now issue #2832), adding tests for the new `success` attribute.

chu11 commented:
re-pushed, fixed a conflict with 557c52f8f1a8b4658ae6a27181e1f044a56d324b and fixed what I hope was a chain-lint issue leading to the builder hanging.

chu11 commented:
rebased & re-pushed, added update for `flux-job` and tests in `t2204-job-info` which I previously forget, added some cleanup fixes, fixed up commits due to the rebase

chu11 commented:
all tests passed!  :-) :-) :-)

chu11 commented:
rebased & re-pushed, realized i accidentally output a `strerror (errno)` on a `json_unpack()` error.

chu11 commented:
with #2865 merged, re do this PR, with the `success_hyphen` attribute removed.

chu11 commented:
rebased & re-pushed with adjustments given #2850 going in

garlick commented on src/modules/job-info/job_state.c:
```diff
@@ -883,6 +883,116 @@ static flux_future_t *state_run_lookup (struct job_state_ctx *jsctx,
+    void *handle;
+
+    if (flux_rpc_get_unpack (f, "{s:s}", "eventlog", &s) < 0) {
+        flux_log_error (ctx->h, "%s: error eventlog for %llu",
```
Elsewhere we've been using the convention of `"%ju", (uintmax_t)id" for formatting job ID's.  Maybe use that instead?

chu11 replied:
ahh, I have not mentally committed the `j` specifier to memory yet (I've only recently got `z` down).  I'll do it in a follow up cleanup PR since I use it all over the place.  And we apparently use it in `flux-job.c` all over the place too.

garlick commented on src/modules/job-info/job_state.c:
```diff
@@ -883,6 +883,116 @@ static flux_future_t *state_run_lookup (struct job_state_ctx *jsctx,
+    return f;
+
+ error:
+    saved_errno = errno;
```
`flux_future_destroy()` doesn't clobber errno so save/restore is unnecessary.

garlick commented on src/modules/job-info/list.c:
```diff
@@ -491,7 +491,7 @@ void list_id_cb (flux_t *h, flux_msg_handler_t *mh,
                     const flux_msg_t *msg, void *arg)
 {
-    if (flux_respond_pack (h, msg, "{s:[s,s,s,s,s,s,s,s,s,s,s,s,s]}",
+    if (flux_respond_pack (h, msg, "{s:[s,s,s,s,s,s,s,s,s,s,s,s,s,s]}",
```
Up to you, but would be less error prone to either return a constant JSON string here, or have a function that builds the array from a constant array of strings by iteratively calling `json_array_append()`?

chu11 replied:
yeah, i've been lazy waiting until I hit 80 chars :-)  I'll add this to the cleanup list for a follow up PR.

chu11 commented:
re-based and re-pushed, only tweaking the 1 @garlick comment above, saving several others for cleanups which I will do in another PR.

chu11 commented:
@Mergifyio refresh
